web/source/keyboards/kmwkeyboards.ts

```diff
@@ -1177,27 +1212,44 @@ namespace com.keyman.keyboards {
                 cmd = cmd + '$';
               }
 
+              languageFound = true;
               break;
             }
           }
+          if (!languageFound) {
+            // TODO: Add error msg that no keyboards are available for lgName
+            console.error('No keyboards are available for '+lgName+'. '+
+              'Does it have another language name?');
+          }
```

Comment:
This should use a pattern similar to the next code block, passing the message within an `Error` via `reject()`.

Reply:
I didn't use `reject()` right now because I wanted the for loop to continue processing languages[].

I think the spec will eventually want these errors merged in with successful keyboard stubs.
> If error, construct a JS error including the string (maybe keyboard info on the failing one)
>  * See if we can put the error in fulfilled. And keep the array of successful keyboards
>  * rejected state, returns array, [ { InternalName: 'gff_mahric', error: 'could not load this keyboard because you can\'t spell'}, { InternalName: 'gff_amharic', lang: 'x', … } ]
>  * Add error to the keyboard [metadata object]

Reply:
Please tailor the comment to more explicitly mention this.  To me, the current comment looks like it's already been fulfilled.  `console.error` is, itself, an error msg.

Something like:

`// TODO:  Construct response array with errors for failed-query keyboards and stubs for successfully-queried keyboards`

---------

Though... in that case... wouldn't it be better to get the return-type of the Promise correct the first time?  Sounds like perhaps an associative-array return type might be best for `addLanguageKeyboards`.  (Basically, `java.util.Map`, but how Javascript does it.)

You'd then want something like:

```
// Roughly equivalent to a declaration of java.util.Map<String, KeyboardStub | Error>
var retMap: {[langName: string]: KeyboardStub | Error = {};

for(language : requested_languages) {
  if(found_stub) {
    retMap[language] = found_stub;
  } else {
    retMap[language] = error;
  }
}

resolve(retMap);
```

Reply:
I've cleaned up the TODO comment. But will address the merged return type later.
From the spec, I'm thinking of adding (on a separate PR)
```
'error': string
```
to KeyboardStub and have the errors appended there (with the corresponding `KI` or `KL`)

Reply:
> From the spec, I'm thinking of adding (on a separate PR)
>
> `'error': string`
>
> to KeyboardStub and have the errors appended there (with the corresponding KI or KL)

I'm **not** a fan of that approach; it pollutes the `KeyboardStub` type.  This is the sort of situation that [discriminated unions](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions) are made for.  This approach is already used in a few different areas of KMW.

Reply:
BTW, I do not plan to approve this PR until this point is addressed, be it within this PR or in a dependent, follow-up PR.  Last thing we want is a "temporary" stop-gap becoming not-temporary for a new API function in a manner that affects its type signature.

Reply:
The current spec says
> If error, construct a JS error including the string (maybe keyboard info on the failing one)
>*  See if we can put the error in fulfilled. And keep the array of successful keyboards
>*  rejected state, returns array, [ { InternalName: 'gff_mahric', error: 'could not load this keyboard because you can\'t spell'}, { InternalName: 'gff_amharic', lang: 'x', … } ]
>*  Add error to the keyboard [metadata object]

If @mcdurdin is fine with it, I can make a separate `ErrorStub` for reporting errors (and provide associated language name or keyboard names). We just need a consensus in the spec

Reply:
See #5340 for my proposed error stub.